### PR TITLE
Story/DFCC-1215: Major enhancement work

### PR DESCRIPTION
**Pull request by jbarnes16503**

### CODE REVIEW

File: DFC.TestAutomation.UI/BrowserType.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿// <copyright file="BrowserType.cs" company="PlaceholderCompany">
+// Copyright (c) PlaceholderCompany. All rights reserved.
+// </copyright>
```
Comment (LJGitHubby) [inline DFC.TestAutomation.UI/BrowserType.cs:3]: Update with relevant information or remove.

File: DFC.TestAutomation.UI/DFC.TestAutomation.UI.csproj [hunk 2]
```diff
@@ -18,9 +18,18 @@
...
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <PlatformTarget>x64</PlatformTarget>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <AdditionalFiles Include="..\stylecop.json" Link="stylecop.json" />
+  </ItemGroup>
+  
   <ItemGroup>
```
Comment (LJGitHubby) [inline DFC.TestAutomation.UI/DFC.TestAutomation.UI.csproj:31]: @jbarnes16503 Is there a way to align these <PackageReference> (include; version) details set within DFC.TestAutomation.UI/DFC.TestAutomation.UI.csproj  to those <dependency> (id; version) details within DFC.TestAutomation.UI.nuspec. i.e. Should they always match? There are additional style cop references here.

File: DFC.TestAutomation.UI/BrowserType.cs (hunk 1 above)
Comment (LJGitHubby) [inline DFC.TestAutomation.UI/BrowserType.cs:3]: Remove or update as appropriate - please check other files.

File: DFC.TestAutomation.UI/DFC.TestAutomation.UI.csproj (hunk 2 above)
Comment (jbarnes16503) [reply DFC.TestAutomation.UI/DFC.TestAutomation.UI.csproj:31]: Style Cop needs to be referenced in the .csproj file in order to highlight styling violations throughout the project. You wouldn't expect to see this reference in the .nuspec. To address your alignment query, we have a ticket in the backlog to remove the .nuspec entirely and go with the .csproj to package the NuGet. This has a few technical challenges, hence why it's in the backlog but this will eliminate the need to align the two files. At eh moment the .csproj isnt used to package up the NuGet.

File: DFC.TestAutomation.UI/BrowserType.cs (hunk 1 above)
Comment (jbarnes16503) [reply DFC.TestAutomation.UI/BrowserType.cs]: Done. I have put National Careers Service in place of PlaceholderCompany

File: DFC.TestAutomation.UI.nuspec [hunk 3]
```diff
@@ -11,9 +11,10 @@
     <tags>UI Automation National Careers Service DFC CompUI</tags>
```
[the remark this answers is not in the document]
Comment (jbarnes16503) [reply DFC.TestAutomation.UI.nuspec:11]: DFC and CompUI has been removed.

File: DFC.TestAutomation.UI/BrowserType.cs (hunk 1 above)
Comment (jbarnes16503) [reply DFC.TestAutomation.UI/BrowserType.cs]: Done. See above.

File: DFC.TestAutomation.UI/Helper/BrowserHelper.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        public BrowserType GetBrowserType()
+        {
+            return this.BrowserIndex[this.BrowserName];
+        }
+
+        /// <summary>
+        /// Assesses whether the current test execution is cloud based. It uses the browser name to identify whether it is cloud based.
+        /// </summary>
+        /// <returns>True in the case where the current test execution is cloud based. False if the current execution is not cloud based.</returns>
+        public bool IsExecutingInTheCloud()
```
Comment (LJGitHubby) [inline DFC.TestAutomation.UI/Helper/BrowserHelper.cs:62]: @jbarnes16503 Think it would reduce maintenance if this function simply checks the `this.BrowserIndex[this.BrowserName]` = BrowserType.BrowserStack ....to remove the need for additional stringcomparisons to be added if extra 'cloud' terms added to the `Dictionary`.

Comment (jbarnes16503) [reply DFC.TestAutomation.UI/Helper/BrowserHelper.cs]: I have renamed this to IsExecutingInBrowserStack() as it is more explicit. I've updated the summary and I have applied the logic you have suggested. Thanks.

File: DFC.TestAutomation.UI/Helper/CommonActionHelper.cs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+            return this.RetryHelper.RetryOnException<string>(Action);
+        }
+
+        /// <summary>
+        /// Gets the text value from all IWebElements with a given locator.
+        /// </summary>
+        /// <param name="locator">The IWebELement locator.</param>
+        /// <returns>The string value.</returns>
+        public string GetTextFromElements(By locator)
```
Comment (LJGitHubby) [inline DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:79]: Does this return text in a suitable structure for use? e.g. how would a request for text from a table be returned? ...or within a <DIV> - would paragraphs be conjoined in a single concatenated string?

Comment (jbarnes16503) [reply DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:79]: It simply returns the text value displayed. I understand your reasons for asking this but I feel like this could be a very deep hole to start digging. We can 'clean up' text but to what end? I'd be more inclined to have string extension methods that handle newline characters or different encoding types and allow the user to do what they want with the text value rather than assuming they want it 'cleaned up'.

I think this goes back to the approach - do it when it's needed. Lets see how we get on and also lets remember that this package is public. So it ought to work across a variety of web applications, not just for the NCS.

File: DFC.TestAutomation.UI/Helper/CommonActionHelper.cs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        }
+
+        /// <summary>
+        /// Assesses whether an IWebElement is present.
+        /// </summary>
+        /// <param name="locator">The IWebElement locator.</param>
+        /// <returns>A value indicating whether the IwebElement is present.</returns>
+        public bool IsElementPresent(By locator)
+        {
+            this.WebDriverWaitHelper.SetImplicitWait(500);
```
Comment (LJGitHubby) [inline DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:109]: @jbarnes16503 Why are we specifically setting the implicitWait to 500ms for these two functions? (Aren't we using a general configuration to set this level within the WebDriverWaitHelper.cs file? `this.WebDriverImplicitWait = new WebDriverWait(webDriver, TimeSpan.FromSeconds(this.TimeoutConfiguration.ImplicitWait));`

Comment (jbarnes16503) [reply DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:109]: We have 2 'is the element present?' methods:

1. WaitForElementToBePresent() in the WebDriverWaitHelper
2. IsElementPresent() a you have indicated above

There are cases where you need the element in order to proceed with your tests in which case you'd use number 1 above. If for example your user journey depends on the presence of an object (a cookie banner for example) then you'd want to simply interrogate the page to see if the object is there. If we didn't reset the timeout and the object was not there, then it would potentially keep looking for the implicit wait length of time. This essentially result in a 60 second delay which wouldn't add any additional value.

So In short think of number 1 being 'wait until its there' and number 2 being 'is it there right now?'. I hope that makes sense.

File: stylecop.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
+﻿{
+  // ACTION REQUIRED: This file was automatically added to your project, but it
+  // will not take effect until additional steps are taken to enable it. See the
+  // following page for additional information:
+  //
+  // https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/EnableConfiguration.md
```
Comment (SimonWren) [inline stylecop.json:6]: If the "additional steps" have been added, should this be removed / updated?

File: DFC.TestAutomation.UI/Helper/CommonActionHelper.cs (hunk 6 above)
Comment (LJGitHubby) [reply DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:109]: I does make sense ...but my question was more about the decision to fix this wait at 500. _Again, happy to leave and refine as appropriate if a need present itself._

File: DFC.TestAutomation.UI.nuspec [hunk 8]
```diff
@@ -2,7 +2,7 @@
 <package >
   <metadata>
     <id>DFC.TestAutomation.UI</id>
-    <version>1.0.0</version>
+    <version>2.0.0</version>
```
[the remark this answers is not in the document]
Comment (jbarnes16503) [reply DFC.TestAutomation.UI.nuspec:5]: I see where you're coming from and I'm fine with either. All I would say is its a bit of a moot point as the version is dictated by the pipeline. This value is overwritten during the pack phase. I think this point needs to be a discussion on how we version control but for the time being it's not used.

File: DFC.TestAutomation.UI/Helper/RetryHelper.cs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RetryHelper"/> class.
+        /// </summary>
+        public RetryHelper()
+        {
+            this.SleepDurations = new TimeSpan[] { TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(3) };
+        }
+
+        private TimeSpan[] SleepDurations { get; set; }
```
[the remark this answers is not in the document]
Comment (jbarnes16503) [reply DFC.TestAutomation.UI/Helper/RetryHelper.cs]: Agreed. Thanks.

I have added this to the test execution settings model so we should expect the user to provide 2 things:

1. The number of retries before throwing an exception
2. The Explicit wait in between retry attempts.

File: stylecop.json (hunk 7 above)
Comment (jbarnes16503) [reply stylecop.json:6]: Ahh good spot. No I think we keep this in as I have raised a bug for SonarCloud and I think this is because the "additional steps" have yet to be carried out. Lets leave the removal of this until we have addressed that?

Comment (jbarnes16503) [reply stylecop.json:6]: https://skillsfundingagency.atlassian.net/browse/DFCC-1291

File: DFC.TestAutomation.UI/Helper/CommonActionHelper.cs (hunk 6 above)
Comment (jbarnes16503) [reply DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:109]: I see. No we don't want to do that as the explicit wait (which is applied globally) is set in the users appsettings.json. So we only want to override this in a handful of places and only in cases where we are looking for immeadiate feedback.

Comment (jbarnes16503) [reply DFC.TestAutomation.UI/Helper/CommonActionHelper.cs:109]: I could have set the value to 0 but 500 is half a second and from experience it's always sensible to leave a little wait time. But yes if we find this isn't working for us we have plenty of options.

**Review by LJGitHubby — APPROVED** [review]
